context/context.go

```diff
@@ -20,140 +20,141 @@
 
 package context
 
-import (
-	"sync"
-)
-
-type dependency struct {
-	closers      []Closer
-	dependencies sync.WaitGroup
-}
+import "sync"
 
-// NB(r): using golang.org/x/net/context is too GC expensive
+// NB(r): using golang.org/x/net/context is too GC expensive.
 type ctx struct {
+	sync.WaitGroup
 	sync.RWMutex
-	pool   contextPool
-	closed bool
-	dep    *dependency
+
+	pool       contextPool
+	done       bool
+	finalizers []Finalizer
 }
 
-// NewContext creates a new context
+// NewContext creates a new context.
 func NewContext() Context {
 	return newPooledContext(nil)
 }
 
-// NewPooledContext returns a new context that is returned to a pool when closed
+// NewPooledContext returns a new context that is returned to a pool when closed.
 func newPooledContext(pool contextPool) Context {
 	return &ctx{pool: pool}
 }
 
-func (c *ctx) ensureDependencies(initClosers bool) {
-	if c.dep == nil {
-		c.dep = &dependency{}
+func (c *ctx) IsClosed() bool {
+	c.RLock()
+	done := c.done
+	c.RUnlock()
+
+	return done
+}
+
+func (c *ctx) RegisterFinalizer(f Finalizer) {
+	if c.Lock(); c.done {
+		c.Unlock()
+		return
 	}
-	if !initClosers || c.dep.closers != nil {
+
+	if c.finalizers != nil {
+		c.finalizers = append(c.finalizers, f)
+		c.Unlock()
 		return
 	}
+
 	if c.pool != nil {
-		c.dep.closers = c.pool.GetClosers()
+		c.finalizers = append(c.pool.GetFinalizers(), f)
 	} else {
-		c.dep.closers = allocateClosers()
+		c.finalizers = append(allocateFinalizers(), f)
 	}
-}
 
-func (c *ctx) RegisterCloser(closer Closer) {
-	c.Lock()
-	if c.closed {
-		c.Unlock()
-		return
-	}
-	c.ensureDependencies(true)
-	c.dep.closers = append(c.dep.closers, closer)
 	c.Unlock()
 }
 
 func (c *ctx) DependsOn(blocker Context) {
 	c.Lock()
-	if !c.closed {
-		c.ensureDependencies(false)
-		c.dep.dependencies.Add(1)
-		blocker.RegisterCloser(c)
+
+	if !c.done {
+		c.Add(1)
+		blocker.RegisterFinalizer(c)
 	}
+
 	c.Unlock()
 }
 
-// OnClose handles a call from another context that was depended upon closing
-func (c *ctx) OnClose() {
-	c.dep.dependencies.Done()
+// Finalize handles a call from another context that was depended upon closing.
+func (c *ctx) Finalize() {
+	c.Done()
 }
 
+type closeMode int
+
+const (
+	closeAsync closeMode = iota
+	closeBlock
+)
+
 func (c *ctx) Close() {
-	c.close(false)
+	c.close(closeAsync)
 }
 
 func (c *ctx) BlockingClose() {
-	c.close(true)
+	c.close(closeBlock)
 }
 
-func (c *ctx) close(blocking bool) {
-	var closers []Closer
-
-	c.Lock()
-	if c.closed {
+func (c *ctx) close(mode closeMode) {
+	if c.Lock(); c.done {
 		c.Unlock()
 		return
 	}
-	c.closed = true
-	if c.dep != nil {
-		closers = c.dep.closers[:]
-	}
+
+	c.done = true
+	finalizers := c.finalizers[:]
```

Comment:
This code is a little bit suspicious to me – do we assume that somewhere at some point we call `ctx.Close(); ctx.Reset()`, i.e. we need to keep another reference to the finalizer slice alive because `c.finalizers` can somehow go away?